rust/lance/src/dataset/sql.rs

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::datafusion::LanceTableProvider;
+use crate::Dataset;
+use arrow_array::RecordBatch;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::prelude::SessionContext;
+use lance_core::Error;
+use snafu::location;
+use std::sync::Arc;
+
+/// Customize the params of dataset's sql API.
+#[derive(Clone, Debug)]
+pub struct SqlOptions {
+    /// the dataset to run the SQL query
+    pub(crate) dataset: Option<Dataset>,
+
+    /// the SQL query to run
+    pub(crate) sql: String,
+
+    /// the name of the table to register in the datafusion context
+    pub(crate) table_name: String,
+
+    /// if true, the query result will include the internal row id
+    pub(crate) row_id: bool,
+
+    /// if true, the query result will include the internal row address
+    pub(crate) row_addr: bool,
+}
+
+impl SqlOptions {
+    pub fn table_name(mut self, table_name: &str) -> Self {
+        self.table_name = table_name.to_string();
+        self
+    }
+
+    pub fn with_row_id(mut self, row_id: bool) -> Self {
+        self.row_id = row_id;
+        self
+    }
+
+    pub fn with_row_addr(mut self, row_addr: bool) -> Self {
+        self.row_addr = row_addr;
+        self
+    }
+
+    pub async fn execute(self) -> lance_core::Result<QueryResult> {
+        let ctx = SessionContext::new();
+        ctx.register_table(
+            self.table_name,
+            Arc::new(LanceTableProvider::new(
+                Arc::new(self.dataset.unwrap()),
+                self.row_id,
+                self.row_addr,
+            )),
+        )?;
+        let df = ctx.sql(&self.sql).await?;
```

Comment:
why not just return df? So user can also do things like `explain` if they want.

Reply:
This is based on Weston's comments:

> We could also return some kind of "query results" object which has into_stream and collect methods. This way users still have the convenience of collect but aren't forced to use it.

IMO, this may be a trade-off. Returning the `df` object would be very free and more flexible. But it requires the users to be very familiar with `datafusion`'s interfaces. Returning a simple object of the query result, users may not lose some flexibility, but they are not confused by many APIs coming from `df`.

Reply:
what if we make query result hold the DataFrame, and offer the related methods to wrap important DataFusion APIs? Because I think at least `explain` is pretty important, this is basically a dry run feature so people can know what will be executed if the execution could be expensive.

Reply:
Sounds reasonable, and that would make the query result to be extendability.